Jdbc Join Pushdown

Thanzeel-Hassan-IBM opened the pull request:
Jdbc Join Pushdown 

This PR is opened for internal review purposes.

Ajas-Mangal commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,607 @@
+
+For performing Jdbc JoinPushdown we required below implementations :
+
+1. Create 2 Optimizers called GroupInnerJoinsByConnector and JdbcJoinPushdown
```
Make this as main or sub heading and use only one optimizer details .

Eg: 1) New optimizer GroupInnerJoinsByConnector
2) New Optimizer JdbcJoinPushdown

Ajas-Mangal commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,607 @@
+
+ Below is the overall process that we have in GroupInnerJoinsByConnector optimizer
+
+![Image 1](RFC-0009-jdbc-join-push-down/in-depth-design-image-1.png)
```
GroupInnerJoinsByConnector processing limited to first image.  Not required predicate pushdown part

Thanzeel-Hassan-IBM replied:
done

Ajas-Mangal commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,607 @@
+
+1. Create 2 Optimizers called GroupInnerJoinsByConnector and JdbcJoinPushdown
+
+For Join pushingdown we need to travers through the JoinNode. We are able to create a new optimizer (GroupInnerJoinsByConnector) which implements PlanOptimizer and another optimizer JdbcJoinPushdown which implements ConnectorPlanOptimizer.
```
Explain below items  here
1) why we need a traversal? 
2) Explain how we managed the tree traversal. How it affect for other queries and datasources.
3) Significants of visitFilter and VisitJoin
4) Explain new data structures MultiJoinNode and ConnectorTableHandleSet and how it is used inside GroupInnerJoinsByConnector
5)  How we did grouping . Which are the conditions to not include in the group.
6) How completed source  flatten 
7) How combine the join tables in GroupInnerJoinsByConnector. Explain Equality and nonEquality criteria creation
8) How recreate the joinnode with grouped tablescannode

Ajas-Mangal commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,607 @@
+
+![Image 1](RFC-0009-jdbc-join-push-down/in-depth-design-image-1.png)
+
+After completing GroupInnerJoinsByConnector optimization we need to invoke predicate pushdown optimizer to recreate join criteria from the filter node of the JoinNode. The detailed implantation of GroupInnerJoinsByConnector optimizer is explained in below sessions. 
```
After GroupInnerJoinsByConnector we should invoke JdbcJoinPushdown and not the predicate pushdown optimizer. Predicate pushdown optimizer may invoke after JdbcJoinPushdown optimizer

Thanzeel-Hassan-IBM replied:
fixed

Ajas-Mangal reviewed (changes requested):
Completed upto first point of low level design

Thanzeel-Hassan-IBM commented:
Will start working on the review comments.

Thanzeel-Hassan-IBM commented:
To do : Add details in Test Plan

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,607 @@
+GroupInnerJoinsByConnector optimizer is a PlanOptimizer which is responsible for flattening the JoinNode and adding the sundered nodes to a data structure called MultiJoinNode. 
+
+GroupInnerJoinsByConnector optimizer will work on MultiJoinNode and will group TableScanNodes based on connector name if the connector supports join pushdown. This optimizer will create a single TableScanNode by using a new data structure called ConnectorTableHandleSet from the grouped TableScanNode. ConnectorTableHandleSet is a set of ConnectorTableHandles which is generated from grouped TableSc ...
+GroupInnerJoinsByConnector optimizer will then work on re-creating join node with updated MultiJoinNode structure. The low level design is available [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#low-level-design)
```
Done

Thanzeel-Hassan-IBM replied:
It's not possible. I tried it.

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,607 @@
+![After GroupInnerJoinsByConnector Optimizer](RFC-0009-jdbc-join-push-down/after_Group_opt.png)  
+
+JdbcJoinPushdown optimizer is a ConnectorPlanOptimizer, specific to jdbc tables and it generate a single JdbcTableHandle from the grouped ConnectorTableHandle. The low level design is available [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#low-level-design)
+
```
It's not possible. I tried it.

Thanzeel-Hassan-IBM replied:
link updated

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,607 @@
+After Predicatepushdown optimizer the flow will invoke existing JdbcComputePushdown optimizer and it will pushdown the overall join criteria to the additional predicates.
+
+After all optimization the PlanNode will pass to the presto-base-jdbc module to create the final join query. The final join query is prepared at the connector level using the Querybuilder. It is explained in the low level design [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#low-level-design).
+
```
Linked.

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,607 @@
+
+After GroupInnerJoinsByConnector optimizer and JdbcJoinPushdown optimizer, we will invoke existing Predicatepushdown optimizer. PredicatePushdown optimizer will pushdown the filter and join criteria to the re-created JoinNode using the overall predicate and overall assignment. 
+
+![Predicatepushdown optimizer](RFC-0009-jdbc-join-push-down/after_predicate.png)
```
done

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,607 @@
+GroupInnerJoinsByConnector optimizer will work on MultiJoinNode and will group TableScanNodes based on connector name if the connector supports join pushdown. This optimizer will create a single TableScanNode by using a new data structure called ConnectorTableHandleSet from the grouped TableScanNode. ConnectorTableHandleSet is a set of ConnectorTableHandles which is generated from grouped TableSc ...
+GroupInnerJoinsByConnector optimizer will then work on re-creating join node with updated MultiJoinNode structure. The low level design is available [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#low-level-design)
+
+![After GroupInnerJoinsByConnector Optimizer](RFC-0009-jdbc-join-push-down/after_Group_opt.png)  
```
done

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,607 @@
+# **RFC0009-jdbc-join-push-down for Presto**
+
+## Jdbc join pushdown in presto
```
done

Thanzeel-Hassan-IBM commented:
Pending :
1. Add details of table alias and why we needed it. 
2. Add more clarity on how the two optimizers are working. It's a bit vague right now.
3. Make the images more clear. Some are not readable.

Ajas-Mangal reviewed (changes requested):
Reviewed upto GroupInnerJoinsByConnector

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,698 @@
+1. The GroupInnerJoinsByConnector uses SimplePlanRewriter methods VisitJoin and VisitFilter to traverse through the nodes. The reason we need to traverse the JoinNode is that we need to identify whether the join query (that is created by presto before Pushdown Optimization is going to happen) is able to be processed by the datasource. For this we traverse all the nodes of the join node and valida ...
+2. Flatten all TableScanNode, filter, outputVariables and assignment to a new data structure called MultiJoinNode. Presto already has an existing data structure called multiJoinNode which is used to flatten Plan nodes into list of source nodes. We are using a similar approach to create multiJoinNode.
+3. Grouping the SourceList of multiJoinNode based on jdbc connector. 
+    3.1. We take each item of SourceList and check if it’s a connector which supports join push down. For this we have introduced a new capability in ConnectorCapabilities named "SUPPORTS_JOIN_PUSHDOWN”. 
```
done

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,698 @@
+    3.3. Once it identifies the connector as pushdown supported, it creates a Map with key as connector name and value as a List of tables which are from the connector.
+    3.4. This ensures that no other connector is affected by this optimiser. Only connectors with Join pushdown capability will be pushed down.
+4. Grouping tables for creating join query - based on JDBC datasource capability [link](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#join-query-pushdown-in-presto-jdbc-datasource)
+    4.1. The grouping of tables happens from the Map which is created above. [Point number 3.3]
```
done

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,698 @@
+After Predicatepushdown optimizer the flow will invoke existing JdbcComputePushdown optimizer and it will pushdown the overall join criteria to the additional predicates.
+
+After all optimization the PlanNode will pass to the presto-base-jdbc module to create the final join query. The final join query is prepared at the connector level using the Querybuilder. It is explained in the low level design [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#create-join-query-in-querybuilder).
+
```
fixed.

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,698 @@
+![GroupInnerJoinsByConnector optimizer](RFC-0009-jdbc-join-push-down/after_GroupInnerJoinsByConnector.png)
+
+JdbcJoinPushdown optimizer is a ConnectorPlanOptimizer, specific to jdbc tables and it generate a single JdbcTableHandle from the grouped ConnectorTableHandle. The low level design is available [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#1-groupinnerjoinsbyconnector-optimizer)
+
```
fixed.

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,698 @@
+
+For performing Jdbc JoinPushdown we required below implementations :
+
+We are going to create a new optimizer (GroupInnerJoinsByConnector) which implements PlanOptimizer and another optimizer (JdbcJoinPushdown) which implements ConnectorPlanOptimizer.
```
added.

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,698 @@
+    3.3. Once it identifies the connector as pushdown supported, it creates a Map with key as connector name and value as a List of tables which are from the connector.
+    3.4. This ensures that no other connector is affected by this optimiser. Only connectors with Join pushdown capability will be pushed down.
+4. Grouping tables for creating join query - based on JDBC datasource capability [link](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#join-query-pushdown-in-presto-jdbc-datasource)
+    4.1. The grouping of tables happens from the Map which is created above. [Point number 3.3]
```
done

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
````diff
@@ -0,0 +1,698 @@
+    }
+    ```
+
+#### 1.2. Load GroupInnerJoinsByConnector optimizer based on session flag
````
The explanation is based on session flag itself. Moved it to Session flag section.
https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#5-session-flag

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,698 @@
+    4.1. The grouping of tables happens from the Map which is created above. [Point number 3.3]
+    4.2. For each item in map, based on connector, we get a list of tables/nodes. Each node is then analysed for join pushdown capability and either added to JoinTables List or added back to rewrittenList (If it can not be pushed down).  
+5. If we are able to create a JoinTables list, then we create a single table scan for that and then add to the rewrittenList.
+    5.1. i.e., if there are 4 tables in JoinTables list against Postgres, then we create a single table scan node with ConnectorHandleSet 
```
Indentation fixed. I think using the image here may not be needed.. I can explain it below with image

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,698 @@
+    3.4. This ensures that no other connector is affected by this optimiser. Only connectors with Join pushdown capability will be pushed down.
+4. Grouping tables for creating join query - based on JDBC datasource capability [link](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#join-query-pushdown-in-presto-jdbc-datasource)
+    4.1. The grouping of tables happens from the Map which is created above. [Point number 3.3]
+    4.2. For each item in map, based on connector, we get a list of tables/nodes. Each node is then analysed for join pushdown capability and either added to JoinTables List or added back to rewrittenList (If it can not be pushed down).  
```
It is added below.. Will link it here.

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
````diff
@@ -0,0 +1,698 @@
+        return immutableEnumSet(NOT_NULL_COLUMN_CONSTRAINT, SUPPORTS_JOIN_PUSHDOWN);
+    }
+    ```
+    3.3. Once it identifies the connector as pushdown supported, it creates a Map with key as connector name and value as a List of tables which are from the connector.
````
Will add this below and link it here.

Thanzeel-Hassan-IBM commented:
Updated two images

To do : 
1. update this image : RFC-0009-jdbc-join-push-down/in-depth-design-image-2.png

Thanzeel-Hassan-IBM commented:
1. Updated RFC-0009-jdbc-join-push-down/in-depth-design-image-2.png
2. Deleted RFC-0009-jdbc-join-push-down/in-depth-design-image-3.png
3. Updated details about JdbcComputePushdown Optimizer.

Ajas-Mangal commented on RFC-0009-jdbc-join-push-down.md:
````diff
@@ -0,0 +1,798 @@
+}
+``` 
+
+Here the sources contain all the tables (irrespective of connector) of that JoinNode, filters contain all the predicate of that JoinNode. Now we have completed the flattening of JoinNode.
````
Correct it
filters contain all the predicate of sundered nodes and  joinCriteriaFilters contains predicate of each JoinNode

Thanzeel-Hassan-IBM replied:
corrected

Ajas-Mangal commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,798 @@
+- 5.3. This rewrittenList is used to create another multiJoinNode (rewrittenMultiJoinNode).
+
+#### 6. Create a joinNode for each sourceList
+- Iterate over the rewrittenMultiJoinNode, for each sourceList, call createLeftDeepJoinTree() method. This creates a joinNode with all the nodes in the sourceList.
```
add link to `Recreate left deep join node from the MultiJoinNode source list `for the details

Thanzeel-Hassan-IBM replied:
added

Ajas-Mangal commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,798 @@
+This assumption by presto is done on the basis that till now all queries done by presto to any database is as a select from a single table.
+
+#### Proposed Solution :
+We can add a field 'table_alias' (table alias) to JdbcTableHandle. This is not the table alias that the user passes in his query.
```
Please get details from latest implementation and add here.

Thanzeel-Hassan-IBM replied:
added

Ajas-Mangal commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,798 @@
+
+If the flag is set ('optimizer-inner-join-pushdown-enabled=true'), then it's optimize method will be invoked.
+If it is not set ('optimizer-inner-join-pushdown-enabled=false' or the flag is not set) then it's optimize method will not be invoked.
+
```
There were one more session flag that we required to restrict non-equijoin pushdown. Please add its details too

Thanzeel-Hassan-IBM replied:
Is that currently implemented ?

Thanzeel-Hassan-IBM replied:
added

Ajas-Mangal reviewed (changes requested):
Some more points to handle

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,798 @@
+GroupInnerJoinsByConnector optimizer is a PlanOptimizer which is responsible for flattening the JoinNode and adding the sundered nodes to a data structure called MultiJoinNode. 
+
+GroupInnerJoinsByConnector optimizer will work on MultiJoinNode and will group TableScanNodes based on connector name if the connector supports join pushdown. This optimizer will create a single TableScanNode by using a new data structure called ConnectorTableHandleSet from the grouped TableScanNode. ConnectorTableHandleSet is a set of ConnectorTableHandles which is generated from grouped TableSc ...
+GroupInnerJoinsByConnector optimizer will then work on re-creating join node with updated MultiJoinNode structure. The low level design is available [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#1-groupinnerjoinsbyconnector-optimizer)
```
fixed.

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
````diff
@@ -0,0 +1,798 @@
+```
+#### 2. Flatten all TableScanNode, filter, outputVariables and assignment to a new data structure called MultiJoinNode
+
+If a JoinNode satisfies all JdbcJoinPushdown requirement, then the very first step is to break the left deep tree structure of JoinNode and create a flattened structure called MultiJoinNode.
````
removed

Thanzeel-Hassan-IBM commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,798 @@
+In GroupInnerJoinsByConnector optimize method we need to invoke Rewriter to rewrite the plannode if it contains JoinNode. JoinNode rewrite is possible by overriding the visitJoin() method of SimplePlanRewriter.
+
+In visitJoin() we should have method to validate JdbcJoinPushdown Conditions that is explained in the document. If the JoinNode satisfies the JdbcJoinPushdown Conditions we should pushdown the Join by creating a single table scan node for that Join. 
+
```
added

Ajas-Mangal reviewed (approved):
It is good now. Please submit for OSS review

josephglerin commented on RFC-0009-jdbc-join-push-down.md:
```diff
@@ -0,0 +1,888 @@
+| 2  | TableScanNode [mydb2_table1, mydb2_table2, mydb2_table3] for DB2 | `select * from db2.db2.mydb2_table1 t3, db2.db2.mydb2_table2 t4, db2.db2.mydb2_table5 t5 where t3.dbthirdtablecolumn = t4.dbfourthtablecolumn and t4.dbfourthtablecolumn = t5.dbfifthtablecolumn` |
+
+
+For performing this jdbc join pushdown,  we need to create two logical optimizers GroupInnerJoinsByConnector and JdbcJoinPushdown. 
```
@Thanzeel-Hassan-IBM 
JdbcJoinPushdown is an connector optimizer

Thanzeel-Hassan-IBM commented:
The PR is merged : https://github.com/prestodb/rfcs/pull/32

closing this as it was opened for internal review.
